### ISSUE TITLE
Segfault when using custom objective with xgb.cv

### ISSUE THREAD

**Issue by khotilov** [dmlc/xgboost#300]
The code below usually segfaults with a message like
```
 *** caught segfault ***
address 0x130, cause 'memory not mapped'

Traceback:
 1: preds - labels
 2: obj(pred, dtrain)
 3: xgb.iter.update(fd$booster, fd$dtrain, i - 1, obj)
 4: xgb.cv(param, dtrain, nthread = 8, nround = 2000, nfold = 10,     obj = logobj)
```
and sometimes it would just hang. It's as if it sometimes happens as soon as one tries to use `labels` produced by `labels <- getinfo(dtrain, "label")`. It also segfaults when I try `nthread=1`, but maybe more rarely. 

OS: RHEL6.5
R: 3.1.2
current xboost

I tried running it within gdb, and it seemed like the issues were within R libraries (I don't have R with debuginfo right now, thus the ??'s):
```
Program received signal SIGSEGV, Segmentation fault.
0x0000003d1bc4c619 in ?? () from /usr/lib64/R/lib/libR.so
(gdb) bt
#0  0x0000003d1bc4c619 in ?? () from /usr/lib64/R/lib/libR.so
#1  0x0000003d1bcc9b39 in ?? () from /usr/lib64/R/lib/libR.so
#2  0x0000003d1bc43cdd in ?? () from /usr/lib64/R/lib/libR.so
#3  0x0000003d1bcd67c9 in Rf_eval () from /usr/lib64/R/lib/libR.so
#4  0x0000003d1bcda117 in ?? () from /usr/lib64/R/lib/libR.so
#5  0x0000003d1bcd666d in Rf_eval () from /usr/lib64/R/lib/libR.so
#6  0x0000003d1bcda34a in ?? () from /usr/lib64/R/lib/libR.so
#7  0x0000003d1bcd666d in Rf_eval () from /usr/lib64/R/lib/libR.so
#8  0x0000003d1bcc86bf in Rf_applyClosure () from /usr/lib64/R/lib/libR.so
#9  0x0000003d1bcd6305 in Rf_eval () from /usr/lib64/R/lib/libR.so
#10 0x0000003d1bcda117 in ?? () from /usr/lib64/R/lib/libR.so
... etc...
```

```R
library(xgboost)

data(agaricus.train)
data(agaricus.test)
data_tr <- agaricus.train$data
data_ts <- agaricus.test$data
label_tr <- agaricus.train$label
label_ts <- agaricus.test$label

# introduce 30% of bad labels
flip_idx <- sample.int(length(label_tr), floor(0.3*length(label_tr)))
label_tr[flip_idx] <- !label_tr[flip_idx]
flip_idx <- sample.int(length(label_ts), floor(0.3*length(label_ts)))
label_ts[flip_idx] <- !label_ts[flip_idx]

dtrain <- xgb.DMatrix(data_tr, label = label_tr)
dtest <- xgb.DMatrix(data_ts, label = label_ts)
watchlist <- list(train = dtrain, eval = dtest)

## The usual built-in objective="binary:logistic" runs fine
#param <- list(max.depth = 2, eta = 0.01, gamma=1., subsample=.5, eval_metric="auc", objective="binary:logistic")
#set.seed(123)
#bst <- xgb.train(param, dtrain, nthread = 8, nround = 1000, watchlist=watchlist)
#set.seed(123)
#bcv <- xgb.cv(param, dtrain, nthread = 8, nround = 1000, nfold=10)

# Custom-implemented Logistic loss objective
logobj <- function(preds, dtrain) {
   labels <- getinfo(dtrain, "label")
   preds <- 1/(1 + exp(-preds))
   #print(str(labels))
   #print(str(preds))
   #print(summary(data.frame(labels,preds)))
   grad <- preds - labels
   #hess <- preds * (1 - preds)
   hess <- pmax(preds * (1 - preds), 1e-16)
   return(list(grad = grad, hess = hess))
}
param <- list(max.depth = 2, eta = 0.01, gamma=1., subsample=.5, eval_metric="auc")
## xgb.train runs fine 
#set.seed(123)
#bst <- xgb.train(param, dtrain, nthread = 8, nround = 1000, watchlist=watchlist, obj=logobj)
# xgb.cv crashes most of times 
set.seed(123)
bcv <- xgb.cv(param, dtrain, nthread = 8, nround = 2000, nfold=10, obj=logobj)
# sometimes need to run it twice to crush
bcv <- xgb.cv(param, dtrain, nthread = 8, nround = 2000, nfold=10, obj=logobj)

```

**Comment by tqchen** [comment]
I pushed a fix on this. Please check if the newest version in master fixed the problem.

Thanks!

**Comment by khotilov** [comment]
Unfortunately, I'm still getting segfaults at the same rate. Do they not happen for you?

I've also stumbled upon this rare outcome again:
```
...
[1250]  train-auc:0.735931+0.001707     test-auc:0.692942+0.017031
Error in preds - labels : non-numeric argument to binary operator
> traceback()
3: obj(pred, dtrain)
2: xgb.iter.update(fd$booster, fd$dtrain, i - 1, obj)
1: xgb.cv(param, dtrain, nthread = 8, nround = 2000, nfold = 10,
       obj = logobj)
```

**Comment by tqchen** [comment]
I could reproduce the effort but the problem on my side was fixed by the recent [commit](https://github.com/dmlc/xgboost/commit/3a534d264dfef77ac0184e464388060207ad56ff)

Can you check if you have compiled and installed against the newest version in the master?

Thanks!

**Comment by khotilov** [comment]
Yes, it does work! No segfaults after over 10K iterations.

I have a few separate R package libraries, and yesterday night, in hurry, I was updating the package in one of them, while running R with another. My apologies...

I'm curious though, why would changing the order of when RNG state is saved into .Random.seed affect the GC? What was the main culprit behind this bug?

Thanks much!

**Comment by tqchen** [comment]
great! It was a mistake I make when adding the PRNG support for xgboost. Basically all the variables allocated should be protected until it was returned. So it could have a chance to get garbage anytime in R runtime when it get UNPROTECTed.

Thanks for providing the case to reveal this! I really appreciate that

**Comment by hetong007** [comment]
Hi @tqchen,  I have just met a similar issue.

I am working with a customized loss function, with `booster = "gblinear"`. The loss function is defined as

```
  quantile.obj.lower = function(preds, dtrain, alpha = lower.quantile) {
    labels = getinfo(dtrain, "label")
    mask = as.numeric(labels > preds)
    grad = (alpha * mask) - ((1.0 - alpha) * (1-mask))
    grad = grad * labels
    hess = rep(1, length(grad))
    return(list(grad = -grad, hess = hess))
  }
```

It is basically doing quantile regression.

The traceback is

```
 *** caught segfault ***
address (nil), cause 'unknown'

Traceback:
 1: obj(pred, dtrain)
 2: xgb.iter.update(bst$handle, dtrain, iteration - 1, obj)
 3: xgb.train(params = params.lower, data = dtrain, nrounds = nrounds,     watchlist = list(train = dtrain, valid = dtest), feval = quant_eval,     maximize = TRUE)
 4: sequential.quantile.cv("../data/sequential.cv.feature.mf.davis.rda",     "../data/sequential.cv.xgb.quantile.2.mf.davis.rda", cutoff = 7,     0.95, 0.05, nrounds = 200, params.upper = params, params.lower = params,     seed = i)
An irrecoverable exception occurred. R is aborting now ...
```

My machine is doing valgrind check slowly, so I may get more information on this problem.

**Comment by hetong007** [comment]
Valgrind caught something during the training process:

```
[184]	train-coverage:0.294489	valid-coverage:0.293961 
[185]	train-coverage:0.294448	valid-coverage:0.293961 
[186]	train-coverage:0.294365	valid-coverage:0.293961 
==12026== Invalid read of size 1
==12026==    at 0x4F4EF81: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4CF36: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7F953: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7FD8F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F800B3: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==  Address 0x18fd7670 is 0 bytes inside a block of size 192,400 free'd
==12026==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12026==    by 0x4F79DF4: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7BC31: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4EA4D91: PutRNGstate (in /usr/lib/R/lib/libR.so)
==12026==    by 0x1431C222: XGDMatrixGetInfo_R (xgboost_R.cc:188)
==12026==    by 0x4F0B4F7: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4BBAA: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4EF7D: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==
==12026== Invalid write of size 1
==12026==    at 0x4F4EF9B: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4CF36: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7F953: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7FD8F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F800B3: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==  Address 0x18fd7670 is 0 bytes inside a block of size 192,400 free'd
==12026==    at 0x4C2BDEC: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12026==    by 0x4F79DF4: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F7BC31: Rf_allocVector3 (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4EA4D91: PutRNGstate (in /usr/lib/R/lib/libR.so)
==12026==    by 0x1431C222: XGDMatrixGetInfo_R (xgboost_R.cc:188)
==12026==    by 0x4F0B4F7: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4BBAA: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4EF7D: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
[truncated]
```

For the full record please visit this gist: https://gist.github.com/hetong007/6b341d934f16f28d3ae60a84c7966219

**Comment by dguleria** [comment]
HI He Tong,

I am facing issues running xgb.cv get error , function not found.

Can you please help me with steps to run xgb model to predict on my 600k
records.

Thanks
Deepak

**Comment by hetong007** [comment]
With a longer execution, `XGBoosterPredict_R` also appeared in the valgrind report. 

After I end the valgrind process (yes it keeps running in valgrind after the above error is reported), the final summary comes like: 

```
[207]	train-coverage:0.952888	valid-coverage:0.946611 
^C
Execution halted
==12026== 
==12026== HEAP SUMMARY:
==12026==     in use at exit: 861,733,801 bytes in 57,695 blocks
==12026==   total heap usage: 1,350,753 allocs, 1,293,060 frees, 65,730,582,407 bytes allocated
==12026== 
==12026== 960 bytes in 3 blocks are possibly lost in loss record 135 of 2,863
==12026==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12026==    by 0x4012E14: allocate_dtv (dl-tls.c:296)
==12026==    by 0x4012E14: _dl_allocate_tls (dl-tls.c:460)
==12026==    by 0x545DD92: allocate_stack (allocatestack.c:589)
==12026==    by 0x545DD92: pthread_create@@GLIBC_2.2.5 (pthread_create.c:500)
==12026==    by 0x6E9E905: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==12026==    by 0x1431B67A: XGDMatrixCreateFromMat_R (xgboost_R.cc:73)
==12026==    by 0x4F0B4F7: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4BBAA: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4EF7D: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4DE9F: ??? (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026==    by 0x4F4B9B2: Rf_eval (in /usr/lib/R/lib/libR.so)
==12026== 
==12026== LEAK SUMMARY:
==12026==    definitely lost: 0 bytes in 0 blocks
==12026==    indirectly lost: 0 bytes in 0 blocks
==12026==      possibly lost: 960 bytes in 3 blocks
==12026==    still reachable: 861,732,841 bytes in 57,692 blocks
==12026==         suppressed: 0 bytes in 0 blocks
==12026== Reachable blocks (those to which a pointer was found) are not shown.
==12026== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==12026== 
==12026== For counts of detected and suppressed errors, rerun with: -v
==12026== ERROR SUMMARY: 48185 errors from 69 contexts (suppressed: 0 from 0)
```

To summarize, the current `XGDMatrixGetInfo_R`, `XGDMatrixNumRow_R`, `XGBoosterPredict_R` and `XGDMatrixCreateFromMat_R` are all possibly related to the segfault. If needed I could provide a piece of code and data, however they are not guaranteed to be reproduced.
